metacat-connector-hive/src/main/java/com/netflix/metacat/connector/hive/sql/HiveConnectorFastTableService.java

```diff
@@ -114,43 +84,67 @@ public boolean exists(final ConnectorRequestContext requestContext, final Qualif
+                        final Table existingTable = getHiveMetacatConverters().fromTableInfo(existingTableInfo);
+                        super.update(requestContext, existingTable, tableInfo);
+                    }
+                } finally {
+                    directSqlTable.unlockIcebergTable(tableId);
+                    log.debug("Unlocked Iceberg table {}", tableName);
+                }
+            } catch (IllegalStateException e) {
+                throw new TablePreconditionFailedException(tableName, e.getMessage());
+            }
         } else {
-            queryBuilder.append(" and location in (");
-            uris.forEach(uri -> {
-                queryBuilder.append("?,");
-                params.add(new SqlParameterValue(Types.VARCHAR, uri));
-            });
-            queryBuilder.deleteCharAt(queryBuilder.length() - 1).append(")");
+            super.update(requestContext, tableInfo);
         }
-        final Map<String, List<QualifiedName>> result = Maps.newHashMap();
-        ResultSetExtractor<Map<String, List<QualifiedName>>> handler = rs -> {
-            while (rs.next()) {
-                final String schemaName = rs.getString("schema_name");
-                final String tableName = rs.getString("table_name");
-                final String uri = rs.getString("location");
-                final List<QualifiedName> names = result.computeIfAbsent(uri, k -> Lists.newArrayList());
-                names.add(QualifiedName.ofTable(this.getCatalogName(), schemaName, tableName));
-            }
-            return result;
-        };
-        try {
-            jdbcTemplate.query(queryBuilder.toString(), params.toArray(), handler);
-        } finally {
-            this.fastServiceMetric.recordTimer(
-                HiveMetrics.TagGetTableNames.getMetricName(), registry.clock().wallTime() - start);
+    }
+
+    private boolean isValidIcebergUpdate(final TableInfo existingTableInfo, final TableInfo newTableInfo) {
+        final Map<String, String> existingMetadata = existingTableInfo.getMetadata();
+        final Map<String, String> newMetadata = newTableInfo.getMetadata();
+        final String existingMetadataLocation = existingMetadata != null
+            ? existingMetadata.get(DirectSqlTable.PARAM_METADATA_LOCATION) : null;
+        final String previousMetadataLocation = newMetadata != null
+            ? newMetadata.get(DirectSqlTable.PARAM_PREVIOUS_METADATA_LOCATION) : null;
+        if (StringUtils.isNotBlank(existingMetadataLocation)
+            && Objects.equals(existingMetadataLocation, previousMetadataLocation)) {
+            return true;
+        } else {
+            log.info("Invalid iceberg table metadata location (expected:{}, given:{})",
+                existingMetadataLocation, previousMetadataLocation);
+            return false;
         }
-        return result;
     }
 
+    private boolean isIcebergTable(final TableInfo tableInfo) {
+        return tableInfo.getMetadata() != null
```

Comment:
Looks to me that this can be used in the existing table check in lockingIcebergTable() function.

Reply:
what do u mean?

Reply:
I mean the logic to check existing table. Is this function doing the same thing?
`try {
             tableType = jdbcTemplate.queryForObject(SQL.TABLE_PARAM_LOCK,
                 new SqlParameterValue[] {new SqlParameterValue(Types.BIGINT, tableId),
                     new SqlParameterValue(Types.VARCHAR, PARAM_TABLE_TYPE), }, String.class);
         } catch (EmptyResultDataAccessException ignored) { }
         if (tableType == null || !ICEBERG_TABLE_TYPE.equals(tableType)) {
             throw new InvalidMetaException(tableName, "Originally table is not of type iceberg", null);
         }`

Reply:
Yeah, we could. I want to keep it outside of the lock logic. I pushed the table_type check based on your comments.